segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -308,23 +336,76 @@ private boolean processWrites(List<Write> toExecute) {
                 return false;
             }
         }
+
         return true;
     }
 
     /**
-     * Gets an ordered list of Writes to execute.
+     * Checks each Write in the given list if it is pointing to a closed WriteLedger. If so, it verifies if the write has
+     * actually been committed (in case we hadn't been able to determine its outcome) and updates the Ledger, if needed.
+     *
+     * @param writes An ordered list of Writes to inspect and update.
+     * @return True if any of the Writes in the given list has been modified (either completed or had its WriteLedger
+     * changed).
      */
-    private List<Write> getWritesToExecute() {
-        // Clean up the write queue of all finished writes that are complete (successfully or failed for good)
-        if (!this.writes.removeFinishedWrites()) {
-            return Collections.emptyList();
+    private boolean handleClosedLedgers(List<Write> writes) {
+        if (writes.size() == 0 || !writes.get(0).getWriteLedger().ledger.isClosed()) {
+            // Nothing to do. We only need to check the first write since, if a Write failed with LedgerClosed, then the
+            // first write must have failed for that reason (a Ledger is closed implies all ledgers before it are closed too).
+            return false;
         }
 
-        // Calculate how much estimated space there is in the current ledger.
-        final long maxTotalSize = this.config.getBkLedgerMaxSize() - getWriteLedger().ledger.getLength();
+        WriteLedger currentLedger = getWriteLedger();
+        Map<Long, Long> lastAddsConfirmed = new HashMap<>();
+        boolean anythingChanged = false;
+        for (Write w : writes) {
+            if (w.isDone() || !w.getWriteLedger().ledger.isClosed()) {
+                continue;
+            }
 
-        // Get the writes to execute from the queue.
-        return this.writes.getWritesToExecute(maxTotalSize);
+            // Write likely failed because of LedgerClosedException. Need to check the LastAddConfirmed for each
+            // involved Ledger and see if the write actually made it through or not.
+            long lac = tryGetLastAddConfirmed(w.getWriteLedger(), lastAddsConfirmed);
+            if (w.getEntryId() >= 0 && w.getEntryId() <= lac) {
```

Comment:
The entry id is given in the response to an add entry, but if `lac < w.getEntryId()`, then we haven't added this entry yet. Are we simply guessing what the entry id will be and setting it initially to the write? It is not hard to guess given that it increases sequentially, but I wonder if this is what we are doing, guessing the entry id for the write ahead of time.